flamingpy/codes/graphs/egraph.py

```diff
@@ -214,3 +217,149 @@ def draw_adj(self, **kwargs):
 
         adj = self.adj_generator(sparse=False)
         return plot_mat_heat_map(adj, **kwargs)
+
+    def add_node(self, node: tuple, **kwargs):
+        """Overwrite add_node function from nx.Graph to raise a warning
+        whenever add_qubit should be used."""
+        if not (self.to_indices is None and self.adj_mat is None):
+            warnings.warn(
+                "EGraph attributes are already populated. To add a"
+                " node, please use the add_qubit function."
+            )
+        super().add_node(node, **kwargs)
+
+    def remove_node(self, node: tuple):
+        """Overwrite remove_node function from nx.Graph to raise a warning
+        whenever remove_node should be used."""
+        if not (self.to_indices is None and self.adj_mat is None):
+            warnings.warn(
+                "EGraph attributes are already populated. To remove a"
+                " node, please use the remove_qubit function."
+            )
+        super().remove_node(node)
+
+    def add_qubit(
+        self,
+        qubit: Union[None, tuple] = None,
+        neighbors: Union[None, list] = None,
+        macro: Union[None, tuple] = None,
+    ) -> None:
+        """Add qubit to EGraph at position with neigbours in
+        existing_neighbours.
+
+        Args:
+            qubit (3D tuple, or None): qubit to add. If it is a 3D tuple,
+            the qubit is added in that position. If qubit is None, the qubit is
+            positioned one unit further than the maximum position in the z
+            direction in position (0,0,z_max + 1).
+
+            neighbours (list[int], list[tuple], or None): neighbors of qubit specified
+            with indices or positions, respectively.
+        """
+
+        # Makes sure that input qubit value and type is supported
+        if isinstance(qubit, tuple):
+            if len(qubit) != 3:
+                raise ValueError(
+                    "The position should be a 3D tuple, but it was given a "
+                    + f"{len(qubit)}D tuple"
+                )
+            if qubit in self:
+                raise ValueError(
+                    f"Could not add qubit because there is already a node in position {qubit}."
+                )
+        elif qubit is None:
+            z_max = max(map(lambda tup: tup[2], self.nodes()))
+            qubit = (0, 0, z_max + 1)
```

Comment:
I actually suggest we add the qubit spatially (e.g. (x_max + 1, 0, 0)). The reason: when you're adding a qubit, this is likely to modify the logical patch rather to extend the code in time.

Reply:
I agree, this makes more sense. Thanks! I've changed this! ✅